### DISCUSSION TITLE
normalize_homography for HomographyWarper

### DISCUSSION THREAD

**Discussion by darkAlert** [kornia/kornia#1783] [Questions]
Hello colleagues,

I have points between two images (A and B) normalized to [0;1).

I compute homography  simply with ` theta, _ = cv2.findHomography(np.array(A_pts), np.array(B_pts))`

Then, to warp image A with cv2.warpPerspective, I first renormalize _theta_ to target size with the following helper function:
```
def rescale_theta(src_size, dst_size, theta):
    ''' Rescales theta (homography) to the target size '''
    src_w, src_h = src_size
    dst_w, dst_h = dst_size
    src_scale = np.array([[dst_w, 0, 0], [0, dst_h, 0], [0, 0, 1]], dtype=np.float64)
    dst_scale_inv = np.array([[1 / src_w, 0, 0], [0, 1 / src_h, 0], [0, 0, 1]], dtype=np.float64)
    scaled_theta = np.matmul(np.matmul(src_scale, theta), dst_scale_inv)

    return scaled_theta

```

unfortunately I don't remember where I found this function but it works for cv2:
```
scaled_theta = rescale_theta(A_size, B_size, theta)      # size = (w,h)
warped_A = cv2.warpPerspective(A, scaled_theta, B_size)
```

And I decided to replace cv2 with kornia. I know that there is difference between `HomographyWarper` and `warp_perspective` / `cv2.warpPerspective` due to normalization. So I tried to just use `kornia.geometry.conversions.normalize_homography` for _theta_ but it didn't work. 

But after some experimentation, I found a working solution:
```
scaled_theta = rescale_theta(A_size, B_size, np.linalg.inv(theta))
norm_theta = normalize_homography(scaled_theta, (A_size[1],A_size[0]), (B_size[1],B_size[0]))

warper = HomographyWarper(B_size[1], B_size[0], normalized_coordinates=True)
warped_A = warper(A, norm_theta)
```
And I couldn't find a solution without using rescale_theta()

Now it works but I'm just wondering why :)  Can anyone help explain this (why do I need to normalize theta twice (with my rescale_theta and then normalize_homography, and also why np.linalg.inv is required before rescale_theta)?

**Comment by edgarriba** [comment]
Hi, not very where your transform matrix comes from, but this rescaling happens already inside `normalize_homography`. The main idea is that opencv warp perspective requires two homography scenarios: 1. Pixel space (0,255) range; 2. Ability to up/downscale image resolution, while the pytorch grid sample requires-1,1 range and is agnostic to the image resolution. The Kornia warp perspective is just a sugar syntax interface for opencv, while homography warper works directly in normalised/pytorch space.

**Reply by edgarriba** [reply]
One quick thing from your code, are you sure that you pass `normalize_homography(b_homo_a, h,w)` ? I see your indexing for sizes as inversed.
